errors/src/commonMain/kotlin/dev/icerock/moko/errors/handler/ExceptionHandlerContext.kt

```diff
@@ -2,51 +2,19 @@
  * Copyright 2020 IceRock MAG Inc. Use of this source code is governed by the Apache 2.0 license.
  */
 
-@file:Suppress("TooGenericExceptionCaught")
-
 package dev.icerock.moko.errors.handler
 
-import dev.icerock.moko.errors.ErrorEventListener
 import dev.icerock.moko.errors.HandlerResult
-import dev.icerock.moko.mvvm.dispatcher.EventsDispatcher
 import kotlin.reflect.KClass
 
-typealias Catcher = (Throwable) -> Boolean
-
-class ExceptionHandlerContext<T> internal constructor(
-    private val eventsDispatcher: EventsDispatcher<ErrorEventListener>,
-    private val block: suspend () -> T
-) {
-    val catchersMap = mutableMapOf<KClass<*>, Catcher>()
-
-    private var finallyBlock: (() -> Unit)? = null
-
-    inline fun <reified E : Throwable> catch(
-        noinline catcher: (E) -> Boolean
-    ): ExceptionHandlerContext<T> {
-        catchersMap[E::class] = catcher as Catcher
-        return this
-    }
-
-    fun finally(block: () -> Unit): ExceptionHandlerContext<T> {
-        finallyBlock = block
-        return this
-    }
-
-    suspend fun execute(): HandlerResult<T, Throwable> {
-        return try {
-            HandlerResult.Success(block())
-        } catch (e: Throwable) {
-            val isHandled = catchersMap[e::class]?.invoke(e)
-            if (isHandled == null || isHandled == false) {
-                eventsDispatcher.dispatchEvent {
-                    showError(e)
-                }
-            }
+interface ExceptionHandlerContext<R> {
+    fun <E : Throwable> catch(clazz: KClass<E>, catcher: (E) -> Boolean): ExceptionHandlerContext<R>
+    fun finally(block: () -> Unit): ExceptionHandlerContext<R>
+    suspend fun execute(): HandlerResult<R, Throwable>
+}
 
-            HandlerResult.Failure(e)
-        } finally {
-            finallyBlock?.invoke()
-        }
-    }
+inline fun <reified E : Throwable> ExceptionHandlerContext<*>.catch(
+    noinline catcher: (E) -> Boolean
+): ExceptionHandlerContext<*> {
+    return catch(E::class, catcher)
 }
```

Comment:
to use inline fun without additional imports we can change `ExceptionHandlerContext` from interface to `abstract class`:
```
abstract class ExceptionHandlerContext<R> {
    abstract fun <E : Throwable> catch(clazz: KClass<E>, catcher: (E) -> Boolean): ExceptionHandlerContext<R>
    abstract fun finally(block: () -> Unit): ExceptionHandlerContext<R>
    abstract suspend fun execute(): HandlerResult<R, Throwable>

    inline fun <reified E : Throwable> catch(
        noinline catcher: (E) -> Boolean
    ): ExceptionHandlerContext<R> {
        return catch(E::class, catcher)
    }
}
```

Reply:
and add factory method to companion object:
```
abstract class ExceptionHandlerContext<R> {
    abstract fun <E : Throwable> catch(clazz: KClass<E>, catcher: (E) -> Boolean): ExceptionHandlerContext<R>
    abstract fun finally(block: () -> Unit): ExceptionHandlerContext<R>
    abstract suspend fun execute(): HandlerResult<R, Throwable>

    inline fun <reified E : Throwable> catch(
        noinline catcher: (E) -> Boolean
    ): ExceptionHandlerContext<R> {
        return catch(E::class, catcher)
    }

    companion object {
        operator fun <T : Any, R> invoke(
            errorPresenter: ErrorPresenter<T>,
            eventsDispatcher: EventsDispatcher<ErrorEventListener<T>>,
            block: suspend () -> R
        ): ExceptionHandlerContext<R> {
            return ExceptionHandlerContextImpl(
                errorPresenter = errorPresenter,
                eventsDispatcher = eventsDispatcher,
                block = block
            )
        }
    }
}
```
after it you can set `ExceptionHandlerContextImpl` is internal, and nobody outside will not see any uncessary classes

Reply:
Reworked.